keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2861 @@
+        configurations are misconfigured. (also via a Condition?).
+      - Add a detailed error in the status of the `ValidatingAdmissionPolicy`.
+      - Trigger the `FailurePolicy` on admission for resources that match the
+        policy configuration.
+  - If the CRD is deleted:
+    - Set state to "misconfigured
+    - Trigger `FailurePolicy` on all admission validations
+    - Add a detailed error in the status of the `ValidatingAdmissionPolicy`.
+
+A partial `spec.match` schema (subset of the full schema) is okay so long as
+only optional fields are omitted. But any unrecognized field in the `spec.match`
+would not be allowed. 
+
+Proposed annotation: 
+
+> Example: `admission.kubernetes.io/is-policy-configuration-definition: "true"`
+> 
+> Used on: CustomResourceDefinition
+> 
+> What a CustomResourceDefinition has the annotation set to "true", the
+> OpenAPIv3 schema of all versions of this resource is modified and then
+> kept-in-sync by a controller to always contain the expected schema fields of
+> admission policy configuration resources.
+
+xref: https://kubernetes.io/docs/reference/labels-annotations-taints/
+
+Pros:
+
+- A single resource is used to configure both the match criteria and
+  configuration params of a policy.
+- Provides a clear marker on every CRD that is used for policy configuration
+  that it is intended for this purpose, in a machine readable way.
+
+Cons:
+
+- API server must check for a wide range of error conditions and define how
+  exactly it handles each of them.
+- If the `spec.match` schema is incorrectly defined, CRD author might not
+  realize it since they need to check the status of the corresponding
+  `ValidatingAdmissionPolicy` for any errors.
+
+###### Alternative: `PolicyConfiguration` kind with config reference
+
+Require two resources:
+
+- A `PolicyConfiguration` resources contains the match criteria and a reference
+  to a configuration resource.
+- A configuration resource that contains only configuration (no match criteria).
+
+Example:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: PolicyConfiguration
+metadata:
+  name: "xyzlimit-scale.example.com"
+  labels:
+    policyConfig: xyzlimit
+spec:
+  match:
+    resourceRules:
+      - apiGroups: ["*"]
+        apiVersions: ["*"]
+        resources: ["*/scale"]
+  config:
+    name: xyzlimit-scale-config.example.com
+```
+
+```yaml
+apiVersion: rules.example.com/v1
+kind: XyzLimit
+metadata:
+  name: "xyzlimit-config.example.com"
+spec:
+  allowed: ["a", "b", "c"]
+  banned: ["x", "y", "z"]
+  xyz:
+    fuzzFactor: 0.8
+    reticulate: true
+```
+
+Note that for policy configurations that require no parameterization, only the
+`PolicyConfiguration` is needed.
+
+The `metadata.labels.policyConfig` label on `PolicyConfiguration` is needed so
+that each `ValidatingAdmissionPolicy` use a label selector to reference the
+policy configurations that configure it.
```

Comment:
Why is this query needed? It needs to be in a field also, not just a label.

Reply:
(apiserver is going to have to load all VAPs and PCs anyway so there is no need to make the query efficient)

Reply:
Good point. Dropping this.